### PR TITLE
feat: expose WarehouseClient method to stream results

### PR DESCRIPTION
**Pull request by ZeRego**
Closes: [#10537](https://github.com/lightdash/lightdash/issues/10537)

### Description:

Expose new WarehouseClient method `streamQuery` and deprecate `runQuery`.
Note that we are still using `runQuery` and all rows loaded in memory. We will incrementally refactor our code to use the new streaming method.

### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

test-api test-backend test-frontend

### CODE REVIEW

File: packages/warehouses/src/warehouseClients/DatabricksWarehouseClient.test.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+                    close: jest.fn(async () => undefined),
+                })),
+                close: jest.fn(async () => undefined),
+            })),
+            close: jest.fn(async () => undefined),
+        })),
+    })),
+}));
+
+describe('DatabricksWarehouseClient', () => {
```
Comment (ZeRego) [inline packages/warehouses/src/warehouseClients/DatabricksWarehouseClient.test.ts:23]: Added a sanity test for the Databricks client.

File: packages/warehouses/src/warehouseClients/WarehouseBaseClient.ts [hunk 2]
```diff
@@ -39,13 +39,40 @@ export default class WarehouseBaseClient<T extends CreateWarehouseCredentials>
...
         throw new Error('Warehouse method not implemented.');
     }
 
+    async runQuery(
+        sql: string,
+        tags?: Record<string, string>,
+        timezone?: string,
+    ) {
+        let fields: WarehouseResults['fields'] = {};
+        const rows: WarehouseResults['rows'] = [];
```
Comment (rephus) [inline packages/warehouses/src/warehouseClients/WarehouseBaseClient.ts:59]: nice wrapping, but I'm wondering if copying the `data.rows` into  a new `rows` means duplicating all the memory usage for each request now

File: packages/warehouses/src/warehouseClients/WarehouseBaseClient.ts [hunk 3]
```diff
@@ -39,13 +39,40 @@ export default class WarehouseBaseClient<T extends CreateWarehouseCredentials>
...
+
+        await this.streamQuery(
+            sql,
+            (data) => {
+                fields = data.fields;
```
Comment (rephus) [inline packages/warehouses/src/warehouseClients/WarehouseBaseClient.ts:64]: so you're overwritting fields with the last chunk of data. Is this compatible with all warehouses ?

File: packages/common/src/types/warehouse.ts [hunk 4]
```diff
@@ -24,14 +29,27 @@ export interface WarehouseClient {
         }[],
     ) => Promise<WarehouseCatalog>;
 
+    streamQuery(
+        query: string,
+        streamCallback: (data: WarehouseResults) => void,
```
Comment (rephus) [inline packages/common/src/types/warehouse.ts:34]: just curious. do we know roughly how big are the chunks per warehouse ? is it random ? is it a fixed number ? are they consistent per warehouse ?

**Review by rephus — COMMENTED** [review]
Code look ok, nice wrapping. 

I'm triggering e2e tests to ensure these requests work for all warehouses

File: packages/warehouses/src/warehouseClients/WarehouseBaseClient.ts (hunk 2 above)
Comment (ZeRego) [reply packages/warehouses/src/warehouseClients/WarehouseBaseClient.ts:59]: It should be the same amount of memory as before as the current code does the same thing. 🤔 
`rows` memory will increase as it gets more entries and `data.rows` only exists while handling that chunk.

File: packages/warehouses/src/warehouseClients/WarehouseBaseClient.ts (hunk 3 above)
Comment (ZeRego) [reply packages/warehouses/src/warehouseClients/WarehouseBaseClient.ts:64]: Yes 👍 I opted to send them in all callback calls to be consistent even if the value is the same. In most warehouse clients we compute the `fields` only once at the start.

File: packages/common/src/types/warehouse.ts (hunk 4 above)
Comment (ZeRego) [reply packages/common/src/types/warehouse.ts:34]: If I'm not mistaken, this is configurable in the warehouse. Example: https://docs.databricks.com/en/structured-streaming/batch-size.html

**Review by rephus — APPROVED** [review]
Tested, everything seems to be working fine 
image


Good job

**Pull request merged**